### PR TITLE
Fix Issue 23617 -  traits(child) compile error need this for something that doesn't need this

### PR DESCRIPTION
**Pull request by RazvanN7**

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

File: compiler/src/dmd/typesem.d [hunk 1]
```diff
@@ -3910,10 +3910,16 @@ Expression dotExp(Type mt, Scope* sc, Expression e, Identifier ident, int flag)
             }
             if (d.needThis() && sc.intypeof != 1)
             {
-                /* Rewrite as:
-                 *  this.d
-                 */
-                if (hasThis(sc))
+                 /* Rewrite as:
+                  * this.d
```
[the remark this answers is not in the document]
Comment (RazvanN7) [reply compiler/src/dmd/typesem.d]: Done.

File: compiler/test/compilable/test23617.d [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+struct Wrapper
+{
+    size_t currentIndex;
+    S[] arrayOfS;
+
+    auto opDispatch(string name, T ...)(T t)
+    {
+        return __traits(child, arrayOfS[this.currentIndex], __traits(getMember, S, name))(t);
```
Comment (adamdruppe) [inline compiler/test/compilable/test23617.d:15]: it should also work without `this` on the `this.currentIndex` so might be worth adding that to the test

```
        __traits(child, arrayOfS[currentIndex], __traits(getMember, S, name))(t); // should work without this
        return __traits(child, arrayOfS[this.currentIndex], __traits(getMember, S, name))(t); // and with
```

**Comment by adamdruppe** [comment]
bleh i missed it before merge, thx day job, but i still suggest confirming both test cases from the bug report work in the PR too

**Comment by RazvanN7** [comment]
@adamdruppe the other case is not fixed by this bug report, it has a different cause. will look into it.

**Pull request merged**